url27db8982cd_leohinterlang_argFace_tgz-pJ8-com_fidelis_argface_ArgSampleJ8/src/com/fidelis/argface/ArgPattern.java

```diff
@@ -0,0 +1,633 @@
+     * 
+     * @param usageNode the node that starts the usage definition
+     * @return {@code true} if a match is found
+     */
+    public boolean matchUsage (ArgNode usageNode) {
+        
+        // Get the number of operands.
+        // If zero, return success.
+        int argCount = nonOptionList.size();
+        if (argCount == 0) {
+            return true;
+        }
+        
+        // Pattern loop.
+        patternMatch = null;
+        ArgNode start = usageNode.getEast();
+        for (String pat = first(start); pat != null; pat = next(start)) {
+            if (pat.isEmpty()) {
+                break;
+            }
+            String text = String.format("%d-%d (%d) %-30.30s %s", patternMin,
+                                  patternMax, argCount, patternSpec(pat),
+                                  pat);
+            trace("pattern", text);
+            if (patternWatch) {
+                if (patternMin <= argCount && argCount <= patternMax) {
+                    System.out.println("Try pattern: " + patternSpec(pat));
+                }
+            }
+            if (match(pat)) {
+                patternMatch = patternSpec(pat);
+                if (patternWatch) {
+                        System.out.println("Pattern match found");
+                }
+                addTargetOperands(pat);
+                return true;
+            } else if (failBase == null) {
+                break;
+            }
+        }
+        help.addProblem(usageNode, "No matching operand pattern");
+        return false;
+    }
+    
+    private String first (ArgNode start) {
+        StringBuilder sb = new StringBuilder();
+        patternList.clear();
+        patternMin = 0;
+        patternMax = 0;
+        String pattern = buildPattern(start, sb);
+        if (patternMax == 0) {
+            patternMax = patternMin;
+        }
+        return pattern;
+    }
+    
+    private String next (ArgNode start) { 
+        StringBuilder sb = new StringBuilder();
+        String pattern = null;
+        nextInit();
+        for (boolean done = false; !done; ) {
+            pattern = rebuildPattern(start, sb);
+            if (pattern == null) {
+                if (lastBase == null) {
+                    return null;
+                }
+                trace("next no pat", lastFail.brief());
+                failBase = lastBase;
+                failNode = lastFail;         
+                sb.setLength(0);
+                nextInit();
+            } else {
+                done = true;
+            }
+        }
+        if (patternMax == 0) {
+            patternMax = patternMin;
+        }
+        return pattern;
+    }
+    
+    private void nextInit () {
+        saveList.clear();
+        saveList.addAll(patternList);
+        patternList.clear();
+        patternMin = 0;
+        patternMax = 0;
+        lastBase = null;
+        lastFail = null;
+        prevIndex = 0;
+    }
+    
+    private boolean match (String pattern) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.